### ISSUE TITLE
Performance issues with Redirect.

### ISSUE THREAD

**Issue by lloydkevin** [patrickdemooij9/SeoToolkit.Umbraco#276]
I've run SQL Profiler and used k6 to generate some load tests.
It appears every HTTP request runs a query to look for Redirects from the `SeoToolkitRedirects` table.

Using k6, my load tests went from 162 RPS to 2382 RPS just be disabling this the package.

**Comment by bielu** [comment]
I also can confirm performance of redirects in this package leave a lot on table.
The first thing which I noticed is that, it takes  quite a long time on each request when profiling my solution:

![Image](https://github.com/user-attachments/assets/c7bfa431-2a90-4b81-bcb1-e9873411646f)

So i did test with it and without and difference was in few hundreds ms!



so I went into thinking what I can do to make it faster, first thing checked:
Sql table have no indexes, so run this query:
```

CREATE UNIQUE INDEX "redirects"
ON [SeoToolkitRedirects] ([OldUrl], [Domain], [NewUrl]) WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, IGNORE_DUP_KEY = OFF, DROP_EXISTING = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
;
```
and rebuilded sql index on table, magic happened time down by quite a lot. @patrickdemooij9 would you maybe consider maybe adding sql indexes on table?

**Comment by patrickdemooij9** [comment]
Hi @lloydkevin & @bielu 

Thank you so much for the report. I hadn't quite noticed this before, but I also haven't had a database with a lot of redirects yet. How many redirects are we talking about here, just so I can do some testing on my own end as well.

I think the index will help for sure, I'll take a look at adding that in so the redirects should be faster!

**Comment by lloydkevin** [comment]
> Hi [@lloydkevin](https://github.com/lloydkevin) & [@bielu](https://github.com/bielu)
> 
> Thank you so much for the report. I hadn't quite noticed this before, but I also haven't had a database with a lot of redirects yet. How many redirects are we talking about here, just so I can do some testing on my own end as well.
> 
> I think the index will help for sure, I'll take a look at adding that in so the redirects should be faster!

I'm not sure an index is the issue. It's not the number of redirects that's the issue.

Granted, this is never a year ago.

One thing Umbraco is great at is caching. If you run load tests on certain pages, after initial load zero queries are run to pull content. Everything is loaded from cache (memory or whatever they're doing).

With this package installed, every single request is now bottlenecked because it has to hit the database to check for redirects.

So an index may help, but why not just read all the redirects and cache them?

Or at least cache them individually if they've already been pulled once.

**Comment by patrickdemooij9** [comment]
> > Hi [@lloydkevin](https://github.com/lloydkevin) & [@bielu](https://github.com/bielu)
> > Thank you so much for the report. I hadn't quite noticed this before, but I also haven't had a database with a lot of redirects yet. How many redirects are we talking about here, just so I can do some testing on my own end as well.
> > I think the index will help for sure, I'll take a look at adding that in so the redirects should be faster!
> 
> I'm not sure an index is the issue. It's not the number of redirects that's the issue.
> 
> Granted, this is never a year ago.
> 
> One thing Umbraco is great at is caching. If you run load tests on certain pages, after initial load zero queries are run to pull content. Everything is loaded from cache (memory or whatever they're doing).
> 
> With this package installed, every single request is now bottlenecked because it has to hit the database to check for redirects.
> 
> So an index may help, but why not just read all the redirects and cache them?
> 
> Or at least cache them individually if they've already been pulled once.

I think index is probably part of the issue. Even if it does know that the redirect should be fetched from the database, then it should be quite quick to be able to get it from the database. 

As for the caching, you are right. I think that is probably the best way to completely fix this issue. The main issue I have is that there could be quite some redirects in that cache that are never being hit, just taking up memory. I once thought about implementing a Bloomfilter in order to determine if the redirect url actually exists or not. If it does, we can get it from the database, but in most cases the filter will be able to filter out the cases that are not present. I think that might be the ideal solution here

**Comment by bielu** [comment]
@patrickdemooij9 can you backport same change to v13 package as well?

**Comment by patrickdemooij9** [comment]
> @patrickdemooij9 can you backport same change to v13 package as well?

Yes! I am finishing up the v16 package this week and then will go straight to backporting some features to v13

**Comment by patrickdemooij9** [comment]
@bielu 
This is now also available in the latest v13 package!
